utils/near-rate-limiter/src/framed_read.rs

```diff
@@ -413,4 +424,92 @@ mod tests {
         assert_eq!(throttle_controller.consume_bandwidth_used(), 1015);
         assert_eq!(throttle_controller.consume_bandwidth_used(), 0);
     }
+
+    #[derive(Default)]
+    pub struct Codec {}
+
+    impl Decoder for Codec {
+        type Item = u8;
+        type Error = std::io::Error;
+
+        fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, std::io::Error> {
+            if buf.is_empty() {
+                // not enough bytes to start decoding
+                return Ok(None);
+            }
+            let res = buf[0];
+            buf.advance(1);
+            Ok(Some(res))
+        }
+    }
+
+    unsafe fn noop_clone(_data: *const ()) -> RawWaker {
+        noop_raw_waker()
+    }
+
+    unsafe fn noop(_data: *const ()) {}
+
+    const NOOP_WAKER_VTABLE: RawWakerVTable = RawWakerVTable::new(noop_clone, noop, noop, noop);
+
+    const fn noop_raw_waker() -> RawWaker {
+        RawWaker::new(null(), &NOOP_WAKER_VTABLE)
+    }
+    pub fn noop_waker() -> Waker {
+        // FIXME: Since 1.46.0 we can use transmute in consts, allowing this function to be const.
+        unsafe { Waker::from_raw(noop_raw_waker()) }
+    }
+
+    #[tokio::test]
+    async fn test_throttled_framed_read() -> Result<(), Box<dyn Error>> {
+        let semaphore = PollSemaphore::new(Arc::new(Semaphore::new(0)));
+        let rate_limiter = ThrottleController::new(semaphore, 1, usize::MAX);
+
+        let listener = TcpListener::bind("127.0.0.1:0").await.expect("TcpListener::bind");
+        let addr = listener.local_addr().expect("addr");
+
+        let client_stream = TcpStream::connect(addr).await?;
+
+        let (mut server_tcp_stream, _socket) = listener.accept().await.unwrap();
+
+        // Write some data.
+        server_tcp_stream.write_all(b"hello world!").await?;
+
+        let (read, _write) = tokio::io::split(client_stream);
+        let semaphore = PollSemaphore::new(Arc::new(Semaphore::new(0)));
+        let mut read =
+            ThrottledFrameRead::new(read, Codec::default(), rate_limiter.clone(), semaphore);
+
+        let waker = noop_waker();
+        let mut context = Context::from_waker(&waker);
+        server_tcp_stream.flush().await?;
+        let start = Instant::now();
+        // Unfortunately, the test times out for some reason.
+        loop {
+            assert!(start.elapsed() < Duration::from_secs(10), "timed out {:?}", start.elapsed());
+
+            match Pin::new(&mut read).poll_next(&mut context) {
+                Poll::Pending => {
+                    // this expected
+                }
+                Poll::Ready(x) => {
+                    assert!(false, "got {:?}", x);
+                    break;
+                }
+            }
```

Comment:
What a conundrum! The bug here is that this loop doesn't contain await. So it never yields to the executor and we get a deadlock. Effectively, we have a cooperative multitasking situation, where task a is blocked on task b, but itself runns a `loop {}`, so task `b` never gets a chance to run.

Reply:
A minimal, hacky fix is to insert `tokio::task::yield_now().await;`. A proper fix is to add a dev-dependency on `tokio-stream` and:

```rust
use tokio_stream::StreamExt;

let msg = read.next().await?;
```

instead of a manual poll-loop.

Reply:
@matklad Thanks, it works!.